Update pointwise concat heuristics

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118453
* #118452

This PR updates the heuristics for lowering to pointwise cat to trigger when we have either a small number of arbitrary pointwise inputs (8) or up to 128 pointwise inputs when they correspond to simple pointwise kernels or data movement.

This originally came from an internal use case which noticed poor codegen: https://fb.workplace.com/groups/1075192433118967/posts/1365770660727808. 

In our initial heuristics for lowering to a masked loads pointwise concat kernel we were conservative with the number of inputs we would allow by setting a maximum of 4.

However, I've noticed that we can much more aggressively fuse to pointwise_concat codegen performantly. 

In the following benchmark I compare foreach and pointwise_cat codegen : https://gist.github.com/eellison/2bf83231f2940d9b9b33eb4721d35e15.

Here is the [csv output](https://gist.github.com/eellison/529da68b326e1d832c26c1dcdb42c313). When there is neither `gelu` applied on prologue or epilogue pointwise concat is faster (this is just the data movement case). Applying gelu on the epilogue does not affect this result.  When you apply gelu on the prologue, then as the # of inputs starts to increase you end up getting register spills with pointwise concat and it gets slower. 

![image](https://github.com/pytorch/pytorch/assets/11477974/0d6612b8-d60f-4984-99eb-9b518cd4af74)



![image](https://github.com/pytorch/pytorch/assets/11477974/4dda3341-68f9-4d1d-8334-67d7196371fb)


When I benchmarked with relu instead of gelu, only as inputs got up to 256 did the pointwise and foreach even out.

![image](https://github.com/pytorch/pytorch/assets/11477974/985418f8-ddb8-47c1-baea-ccd9de72cd7f)



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @bertmaher 


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

eellison committed:
Remove # input limit to pointwise cat

[ghstack-poisoned]

eellison committed:
Update on "Remove # input limit to pointwise cat"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Remove # input limit to pointwise cat"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Remove # input limit to pointwise cat"


This originally came from an internal use case which noticed poor codegen: https://fb.workplace.com/groups/1075192433118967/posts/1365770660727808. 

In our initial heuristics for lowering to a pointwise concat kernel which uses masked loads we were conservative with the number of inputs we would allow lowering to by setting a maximum of 4.  However in doing additional benchmarking I don't see any regression as we bump this up. 

When comparing a simple benchmark of concatting a number of inputs together, either with or without a gelu applied before/after pointwise fusion scales well and outperforms for each fusion:
```
def func(input, gelu_prologue=False, gelu_epilogue=False):
    if gelu_prologue:
        input = [torch.nn.functional.gelu(val) for val in input]
    a = torch.stack(input, dim=2).view(-1, 416)
    if gelu_epilogue:
        a = torch.nn.functional.gelu(a)
    return a
 ```
 
 The sizes there were taken from internal. Full benchmark here : https://gist.github.com/eellison/2b7abbc578d1819ba7938d3f997da168 

This also holds when inputting smaller tensors with a larger number (WIP):



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Update pointwise concat heuristics"


This PR updates the heuristics for when we lower to pointwise cat to trigger when we have either a small number of arbitrary pointwise inputs (8) or up to 128 pointwise inputs when they correspond to simple pointwise kernels or data movement.

This originally came from an internal use case which noticed poor codegen: https://fb.workplace.com/groups/1075192433118967/posts/1365770660727808. 

Our initial heuristics for lowering to a pointwise concat kernel which uses masked loads we were conservative with the number of inputs we would allow lowering to by setting a maximum of 4.

However, I've noticed that we can much more aggressively fuse to pointwise_concat codegen performantly. 

In the following benchmark I compares foreach and pointwise_cat codegen : https://gist.github.com/eellison/2bf83231f2940d9b9b33eb4721d35e15.

Here is the [csv output](https://gist.github.com/eellison/529da68b326e1d832c26c1dcdb42c313)(https://gist.github.com/eellison/529da68b326e1d832c26c1dcdb42c313). When there is neither `gelu` applied on prologue or epilogue pointwise concat is faster (this is just the data movement case). Applying gelu on the epilogue does not affect this result.  When you apply gelu on the prologue, then as the # of inputs starts to increase you end up getting register spills with pointwise concat and it gets slower. 

![image](https://github.com/pytorch/pytorch/assets/11477974/0d6612b8-d60f-4984-99eb-9b518cd4af74)



![image](https://github.com/pytorch/pytorch/assets/11477974/4dda3341-68f9-4d1d-8334-67d7196371fb)


When I benchmarked with relu instead of gelu, only as inputs got up to 256 did the pointwise and foreach even out.

![image](https://github.com/pytorch/pytorch/assets/11477974/985418f8-ddb8-47c1-baea-ccd9de72cd7f)



cc bertmaher 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Update pointwise concat heuristics"


This PR updates the heuristics for when we lower to pointwise cat to trigger when we have either a small number of arbitrary pointwise inputs (8) or up to 128 pointwise inputs when they correspond to simple pointwise kernels or data movement.

This originally came from an internal use case which noticed poor codegen: https://fb.workplace.com/groups/1075192433118967/posts/1365770660727808. 

Our initial heuristics for lowering to a pointwise concat kernel which uses masked loads we were conservative with the number of inputs we would allow lowering to by setting a maximum of 4.

However, I've noticed that we can much more aggressively fuse to pointwise_concat codegen performantly. 

In the following benchmark I compares foreach and pointwise_cat codegen : https://gist.github.com/eellison/2bf83231f2940d9b9b33eb4721d35e15.

Here is the [csv output](https://gist.github.com/eellison/529da68b326e1d832c26c1dcdb42c313)(https://gist.github.com/eellison/529da68b326e1d832c26c1dcdb42c313). When there is neither `gelu` applied on prologue or epilogue pointwise concat is faster (this is just the data movement case). Applying gelu on the epilogue does not affect this result.  When you apply gelu on the prologue, then as the # of inputs starts to increase you end up getting register spills with pointwise concat and it gets slower. 

![image](https://github.com/pytorch/pytorch/assets/11477974/0d6612b8-d60f-4984-99eb-9b518cd4af74)



![image](https://github.com/pytorch/pytorch/assets/11477974/4dda3341-68f9-4d1d-8334-67d7196371fb)


When I benchmarked with relu instead of gelu, only as inputs got up to 256 did the pointwise and foreach even out.

![image](https://github.com/pytorch/pytorch/assets/11477974/985418f8-ddb8-47c1-baea-ccd9de72cd7f)



cc bertmaher 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison committed:
Update on "Update pointwise concat heuristics"


This PR updates the heuristics for lowering to pointwise cat to trigger when we have either a small number of arbitrary pointwise inputs (8) or up to 128 pointwise inputs when they correspond to simple pointwise kernels or data movement.

This originally came from an internal use case which noticed poor codegen: https://fb.workplace.com/groups/1075192433118967/posts/1365770660727808. 

In our initial heuristics for lowering to a masked loads pointwise concat kernel we were conservative with the number of inputs we would allow by setting a maximum of 4.

However, I've noticed that we can much more aggressively fuse to pointwise_concat codegen performantly. 

In the following benchmark I compare foreach and pointwise_cat codegen : https://gist.github.com/eellison/2bf83231f2940d9b9b33eb4721d35e15.

Here is the [csv output](https://gist.github.com/eellison/529da68b326e1d832c26c1dcdb42c313). When there is neither `gelu` applied on prologue or epilogue pointwise concat is faster (this is just the data movement case). Applying gelu on the epilogue does not affect this result.  When you apply gelu on the prologue, then as the # of inputs starts to increase you end up getting register spills with pointwise concat and it gets slower. 

![image](https://github.com/pytorch/pytorch/assets/11477974/0d6612b8-d60f-4984-99eb-9b518cd4af74)



![image](https://github.com/pytorch/pytorch/assets/11477974/4dda3341-68f9-4d1d-8334-67d7196371fb)


When I benchmarked with relu instead of gelu, only as inputs got up to 256 did the pointwise and foreach even out.

![image](https://github.com/pytorch/pytorch/assets/11477974/985418f8-ddb8-47c1-baea-ccd9de72cd7f)



cc bertmaher 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 commented on torch/_inductor/lowering.py:
```diff
@@ -1126,10 +1122,33 @@ def can_fuse_reduction(t):
+
+        # this will correspond to a direct memory read
+        if not isinstance(x, ir.Pointwise):
+            return 0
```
Q: what if x is an reduction. i.e., we cat a few reductions together. We would under-estimate the op_count here.  (A bit picky since this may not appear in real model much)

eellison replied:
In prior PR I disable inputs with reduction reads

shunting314 commented on torch/_inductor/lowering.py:
```diff
@@ -1035,17 +1035,14 @@ def inner_fn(idx):
-
-        return get_masked_val(0)
+        next_val = masked_loads[-1]
+        for i in range((len(inputs)) - 1, -1, -1):
```
This can be optimized a little bit by using:
```
for i in range((len(inputs)) - 2, -1, -1):
```
?

although I like binary search more but that may be not trivial for codegen..

eellison committed:
Update on "Update pointwise concat heuristics"


This PR updates the heuristics for lowering to pointwise cat to trigger when we have either a small number of arbitrary pointwise inputs (8) or up to 128 pointwise inputs when they correspond to simple pointwise kernels or data movement.

This originally came from an internal use case which noticed poor codegen: https://fb.workplace.com/groups/1075192433118967/posts/1365770660727808. 

In our initial heuristics for lowering to a masked loads pointwise concat kernel we were conservative with the number of inputs we would allow by setting a maximum of 4.

However, I've noticed that we can much more aggressively fuse to pointwise_concat codegen performantly. 

In the following benchmark I compare foreach and pointwise_cat codegen : https://gist.github.com/eellison/2bf83231f2940d9b9b33eb4721d35e15.

Here is the [csv output](https://gist.github.com/eellison/529da68b326e1d832c26c1dcdb42c313). When there is neither `gelu` applied on prologue or epilogue pointwise concat is faster (this is just the data movement case). Applying gelu on the epilogue does not affect this result.  When you apply gelu on the prologue, then as the # of inputs starts to increase you end up getting register spills with pointwise concat and it gets slower. 

![image](https://github.com/pytorch/pytorch/assets/11477974/0d6612b8-d60f-4984-99eb-9b518cd4af74)



![image](https://github.com/pytorch/pytorch/assets/11477974/4dda3341-68f9-4d1d-8334-67d7196371fb)


When I benchmarked with relu instead of gelu, only as inputs got up to 256 did the pointwise and foreach even out.

![image](https://github.com/pytorch/pytorch/assets/11477974/985418f8-ddb8-47c1-baea-ccd9de72cd7f)



cc bertmaher 


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

eellison commented:
@pytorchbot merge
